enclave-runner/src/usercalls/mod.rs

```diff
@@ -793,12 +865,98 @@ impl EnclaveState {
         num_of_worker_threads: usize,
         start_work: Work,
     ) -> StdResult<(u64, u64), EnclaveAbort<EnclavePanic>> {
+
         fn create_worker_threads(
+            enclave: Arc<EnclaveState>,
             num_of_worker_threads: usize,
             work_receiver: crossbeam::crossbeam_channel::Receiver<Work>,
             io_queue_send: tokio::sync::mpsc::UnboundedSender<UsercallSendData>,
+            async_ret_rx: crossbeam::crossbeam_channel::Receiver<(Return, usize)>
         ) -> Vec<JoinHandle<()>> {
             let mut thread_handles = vec![];
+            let writer_waker = Arc::new(tokio::prelude::task::AtomicTask::new());
+            let reader_waker = Arc::new(tokio::prelude::task::AtomicTask::new());
+
+            //loop through async_queue
+            let enclave_clone = enclave.clone();
+            let mut io_queue_send_clone = io_queue_send.clone();
+            let reader_waker_clone = reader_waker.clone();
+            let writer_waker_clone = writer_waker.clone();
+            thread::spawn(move || {
+
+                tokio::runtime::current_thread::block_on_all(tokio::prelude::future::poll_fn(|| -> Poll<(),()> {
+                    loop {
+                        let val = unsafe {enclave_clone.async_queue_usercall.recv()};
+                        match val {
+                            None => {
+                                // sleep
+                                // retry after waking up
+                                reader_waker_clone.register();
```

Comment:
How is this supposed to be woken up when the enclave sends a usercall?

Reply:
:man_facepalming: 
I have changed the code for both of the threads to be woken up on any synchronous usercall according to the following
https://edp.fortanix.com/docs/api/fortanix_sgx_abi/async/index.html#enclaveuserspace-synchronization